File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;
struct cmp{
    bool operator()(vector<int> a, vector<int> b){
        return a[1] > b[1];
    }
};

int solution(vector<vector<int>> jobs) {
    int answer = 0, i = 0, time = 0;
    priority_queue<vector<int>, vector<vector<int>>, cmp> pq;
    sort(jobs.begin(), jobs.end());
    while(i < jobs.size() || !pq.empty()){
        if(i < jobs.size() && time >= jobs[i][0]){
            pq.push(jobs[i++]);
            continue;
        }
        if(!pq.empty()) {
            time += pq.top()[1];
            int exec = time - pq.top()[0];
            answer += exec;
            pq.pop();
        }
        else
            time = jobs[i][0];
    }
    return answer/jobs.size();
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int answer = 0, i =0;
        int time = 0;
        PriorityQueue<ArrayList<Integer>> pq = new PriorityQueue<>((a, b) -> a.get(1) - b.get(1));
        Arrays.sort(jobs, (a, b) -> a[0] - b[0]);
        while(i < jobs.length || !pq.isEmpty()){
            if(i < jobs.length && time >= jobs[i][0]){
                ArrayList<Integer> l = new ArrayList<>();
                l.add(jobs[i][0]);
                l.add(jobs[i][1]);
                pq.add(l);
                i++;
                continue;
            }
            if(!pq.isEmpty()){
                time += pq.peek().get(1);
                int exec = time - pq.peek().get(0);
                answer += exec;
                pq.poll();
            }
            else time = jobs[i][0];
        }
        return answer/jobs.length;
    }
}